tests/ec-keygen.c

```diff
@@ -0,0 +1,229 @@
+		goto err;
+	}
+	if (EVP_DigestVerifyFinal(mdctx, sig, slen) == 1) {
+		printf("Verify success\n");
+		retval = 0;
+		goto err;
+	} else {
+		display_openssl_errors(__LINE__);
+		printf("Verify fail\n");
+		retval = -10;
+		goto err;
+	}
+
+err:
+	if (sig)
+		OPENSSL_free(sig);
+	if (mdctx)
+		EVP_MD_CTX_destroy(mdctx);
+	return retval;
+}
+
+int main(int argc, char* argv[])
+{
+	int ret = EXIT_FAILURE, res;
+	ENGINE* engine = NULL;
+	const char *efile, *module;
+	char *key_pass;
+	EVP_PKEY *ecpb = NULL;
+	EVP_PKEY *ecpr = NULL;
+	PKCS11_EC_KGEN ec = {
+		.curve = "P-256"
+	};
+	PKCS11_params params = {
+		.sensitive = 1,
+		.extractable = 0,
+	};
+	PKCS11_KGEN_ATTRS eckg = {
+		.type = EVP_PKEY_EC,
+		.kgen.ec = &ec,
+		.token_label = NULL,
+		.key_label = NULL,
+		.key_id = (const unsigned char *)"\x22\x33",
+		.id_len = 2,
+		.key_params = &params,
+	};
+
+	if (argc < 5) {
+		printf("Too few arguments\n");
+		printf("%s [TOKEN1] [KEY-LABEL] [PIN] [CONF] [module]\n", argv[0]);
+		goto cleanup;
+	}
+	eckg.token_label = argv[1];
+	eckg.key_label = argv[2];
+	key_pass = argv[3];
+	efile = argv[4];
+	module = argv[5];
+
+	res = CONF_modules_load_file(efile, "engines", 0);
+	if (res <= 0) {
+		printf("cannot load %s\n", efile);
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+
+	ENGINE_add_conf_module();
+#if OPENSSL_VERSION_NUMBER>=0x10100000
+	OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS \
+		| OPENSSL_INIT_ADD_ALL_DIGESTS \
+		| OPENSSL_INIT_LOAD_CONFIG, NULL);
+#else
+	OpenSSL_add_all_algorithms();
+	OpenSSL_add_all_digests();
+	ERR_load_crypto_strings();
+#endif
+	ERR_clear_error();
+
+	ENGINE_load_builtin_engines();
+	engine = ENGINE_by_id("pkcs11");
+	if (engine == NULL) {
+		printf("Could not get engine\n");
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+
+	if (!ENGINE_ctrl_cmd_string(engine, "PIN", key_pass, 0)) {
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+	if (!ENGINE_ctrl_cmd_string(engine, "DEBUG_LEVEL", "7", 0)) {
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+	if (module) {
+		if (!ENGINE_ctrl_cmd_string(engine, "MODULE_PATH", module, 0)) {
+			display_openssl_errors(__LINE__);
+			goto cleanup;
+		}
+	}
+	if (!ENGINE_init(engine)) {
+		printf("Could not initialize engine\n");
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+	/*
+	 * ENGINE_init() returned a functional reference, so free the structural
+	 * reference from ENGINE_by_id().
+	 */
+	ENGINE_free(engine);
+
+	/*
+	 * EC key generation test
+	 */
+	if (!ENGINE_ctrl_cmd(engine, "KEYGEN", 0, &eckg, NULL, 1)) {
+		printf("Could not generate ECC keys\n");
+		goto cleanup;
+	}
+	printf("ECC keys generated\n");
+
+	ecpb = ENGINE_load_public_key(engine, "2233", NULL, NULL);
+	ecpr = ENGINE_load_private_key(engine, "2233", NULL, NULL);
+	if ((ret = sign_verify_test(ecpr, ecpb)) < 0) {
+		printf("ECC Sign-verify failed with err code: %d\n", ret);
+		goto cleanup;
+	}
+	printf("ECC Sign-verify success\n");
```

Comment:
Formally, it should be EC keys instead of ECC keys, as EC refers specifically to Elliptic Curve keys, while ECC refers to the broader field of Elliptic Curve Cryptography.